SQS: Add send a message tab

ranyoo2367 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description
<!--- Describe your changes in detail -->
Added a tab for sending message. 
- FIFO Queues:
   - Asks for message body, deduplication ID, and group ID.
   - If input fields are empty upon pressing send, error messages are displayed.
   - Context help added for deduplication and group IDs.
- Standard Queues:
   - Asks for message body.
   - If message body is empty upon pressing send, error message is displayed.
- A confirmation with the message ID is displayed if message is sent successfully. If it fails to send the message, an error message is displayed.

## Testing
<!--- Please describe in detail how you tested your changes -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- Tested form validation to make sure error messages are displayed if any of the message body, deduplication ID, or group ID fields are empty.
- Tested confirmation/failure of sending message.

## Screenshots (if appropriate)
FIFO:
![FifoEdit](https://user-images.githubusercontent.com/30319724/89362613-4b9c5400-d69c-11ea-93b5-84caee1c53b9.gif)

Standard:
<img width="767" alt="Screen Shot 2020-08-04 at 9 57 28 PM" src="https://user-images.githubusercontent.com/30319724/89363140-7fc44480-d69d-11ea-9f28-341a17ff504c.png">

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [ ] Local run of `gradlew check` succeeds
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [ ] All new and existing tests passed

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

hunterwerlla commented:
If I make the message too long, the box can't scroll and the send button dissapears
<img width="1385" alt="Screen Shot 2020-08-03 at 2 47 31 PM" src="https://user-images.githubusercontent.com/11964782/89230551-5cfc3800-d598-11ea-9a76-9925bc051b19.png">

hunterwerlla commented:
I don't think there needs to be a separator in this menu:
<img width="184" alt="Screen Shot 2020-08-03 at 2 55 23 PM" src="https://user-images.githubusercontent.com/11964782/89231084-76ea4a80-d599-11ea-89e1-7bea6987f608.png">

hunterwerlla commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePane.kt:
```diff
@@ -2,8 +2,119 @@
+            emptyGroupLabel.isVisible = groupId.text.isEmpty()
+        }
+
+        return !(emptyBodyLabel.isVisible || emptyDeduplicationLabel.isVisible || emptyGroupLabel.isVisible)
```
This check should be based on what isVisible is based on, not the state of isVisible. You can save the isEmpty checks to booleans then reuse them

hunterwerlla commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePane.kt:
```diff
@@ -2,8 +2,119 @@
+) : CoroutineScope by ApplicationThreadPoolScope("SendMessagePane") {
     lateinit var component: JPanel
+    lateinit var inputText: JBTextArea
+    lateinit var fifoComponent: JPanel
```
`fifoComponent` does not really explain what it is. Maybe `fifoFields` or `fifoArguments` or something to that effect

hunterwerlla commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePane.form:
```diff
@@ -1,12 +1,155 @@
+          <text value="Message Body:"/>
+        </properties>
+      </component>
+      <grid id="64b6c" binding="fifoComponent" layout-manager="GridLayoutManager" row-count="4" column-count="4" same-size-horizontally="false" same-size-vertically="false" hgap="-1" vgap="-1">
```
consider making fifoComponent its own form, it has as many components as the other form, and can help encapsulate the logic to have less checks in the main form.

hunterwerlla commented:
FIFO queues by default have a tiny message window:
<img width="1392" alt="Screen Shot 2020-08-03 at 3 08 10 PM" src="https://user-images.githubusercontent.com/11964782/89231954-3f7c9d80-d59b-11ea-8adc-3e03fa709b2a.png">

hunterwerlla commented:
Sending a FIFO message with the same deduplication ID and group ID fails silently

hunterwerlla commented:
A general UX thing, as soon as the deduplication/group ID box has text in it, we can hide the messages again

hunterwerlla commented:
If I put in a dedupliction ID that's too long, I get a generic "Failed to send message", we should be able to bubble up this error

ranyoo2367 commented:
New set of changes:
- Fifo components are in a separate form.
- Replaced `PersistentStateComponent` with `PropertiesComponent`
- Added listeners for components
- Added more specific validation errors
- UI changes to enable scrolling for text area

hunterwerlla commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePane.form:
```diff
@@ -1,12 +1,90 @@
+          <grid row="4" column="3" row-span="1" col-span="1" vsize-policy="0" hsize-policy="0" anchor="2" fill="1" indent="0" use-parent-layout="false"/>
+        </constraints>
+        <properties>
+          <text value="Send"/>
```
Needs to be localizable

hunterwerlla commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePane.form:
```diff
@@ -1,12 +1,90 @@
+          <grid row="4" column="4" row-span="1" col-span="1" vsize-policy="0" hsize-policy="0" anchor="2" fill="1" indent="0" use-parent-layout="false"/>
+        </constraints>
+        <properties>
+          <text value="Clear"/>
```
Ditto

hunterwerlla commented on jetbrains-core/tst/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePaneTest.kt:
```diff
@@ -0,0 +1,172 @@
+            runBlocking { sendMessage() }
+        }
+
+        assertTrue { standardPane.bodyErrorLabel.isVisible }
```
Ah I didn't notice this before, but we don't use kotlin test asserts (I need to add that to the import ktlint rule), we use the asssertJ versions

hunterwerlla commented on jetbrains-core/tst/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePaneTest.kt:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    @Test
+    fun `Too long ID fails to send for fifo`() {
```
nit: `ID too long`

hunterwerlla commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/FifoPanel.kt:
```diff
@@ -0,0 +1,97 @@
+        var groupIsValid = true
+
+        if (deduplicationId.text.length > MAX_LENGTH_OF_FIFO_ID) {
+            deduplicationIsValid = false
```
Add a message for this then put the next if in an `else if`?

ranyoo2367 replied:
The message should be taken care of in the document listener, but I can explicitly add it again.

hunterwerlla commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/FifoPanel.kt:
```diff
@@ -0,0 +1,97 @@
+            deduplicationErrorLabel.isVisible = true
+            deduplicationIsValid = false
+        }
+        if (groupId.text.length > MAX_LENGTH_OF_FIFO_ID) {
```
Ditto
